sympy/geometry/point.py

```diff
@@ -261,59 +296,127 @@ def affine_rank(*points):
+
+        >>> p(Point(1, 2, 0), dimension=2, strict=True)
+        Traceback (most recent call last):
+        ...
+        ValueError: Point has wrong number of dimensions
+
+        >>> p((1, 2, 0), dimension=4, strict=True)
+        Traceback (most recent call last):
+        ...
+        ValueError: wrong number of dimensions
+        """
+        if dimension is not None:
+            dimension = as_int(dimension)
+        was_point = False
         if isinstance(a, Point):
             if dimension and a.ambient_dimension != dimension:
-                raise ValueError("Point {} must be of dimension {}.".format(a, dimension))
-            return a
-
-        # any interable that is traversed in a canonical way
-        # (which excludes things like FiniteSets) can be turned
-        # into a point.  This check can be made more general, but for now
-        # exclude known bad objects
-        if iterable(a) and not isinstance(a, (FiniteSet, set, dict)):
-            # in the special case of 2 and 3 dimensions, let the constructor
-            # ensure the dimension is correct.  In particular, the convention
-            # for Point3D is to allow being called with only two arguments
-            # and silently add a zero as the last coordinate.
-            if dimension == 2:
-                return Point.pointify(Point2D(a, **kwargs))
-            elif dimension == 3:
-                return Point.pointify(Point3D(a, **kwargs))
-            return Point.pointify(Point(a, **kwargs), dimension)
-
-        if ensure_geometry_entity and not isinstance(a, GeometryEntity):
-            raise TypeError("{} must be a GeometryEntity".format(a))
-        if ensure_point and not isinstance(a, Point):
-            raise TypeError("{} must be a Point".format(a))
-
+                if strict:
+                    raise ValueError('wrong number of dimensions')
+                was_point = True
+                a = a.args
+            else:
+                return a  # if Point or if dimension not given
+        elif isinstance(a, GeometryEntity):
+            raise TypeError('expecting Point, not {}'.format(func_name(a)))
+        elif not is_sequence(a):
+            raise TypeError(filldedent('''
+                expecting Point or sequence of 2 or
+                more coordinates, not `{}`'.format(a))
+                '''))
+        else:
+            a = Tuple(*a)
+        if dimension:
+            if len(a) != dimension:
+                if strict:
+                    raise ValueError('wrong number of dimensions')
+                if any(i for i in a[dimension:]):
+                    raise ValueError('too many nonzero dimensions')
+                if len(a) > dimension:
+                    a = a[:dimension]  # strip 0s
+                else:
+                    a += (S.Zero,)*(dimension - len(a))  # append zeros
+        if was_point:
+            a = Point(*a, evaluate=False)
+        elif ensure_point:
+            a = Point(*a)
```

Comment:
That's what I was thinking. It just didn't have the "do one thing well" design. See the updated `Point.__new__`.